### DISCUSSION TITLE
Reducing the role of RGB Node

### DISCUSSION THREAD

**Discussion by dr-orlovsky** [RGB-WG/rgb#49] [Ideas]
Orignaly, RGB Node was created with the idea of performing P2P communications for client-side-validated data. However, with the later distinction of data transfer from the core of RGB protocol, and support for multiple data transfer solutions (LN/Storm, relays etc) the need in the dedicated node got much reduced.

Currently, to do nearly anything related to RGB, like validate a consignment or pepare a payment, one needs to run RGB Node, which manages stash. This puts high resource requirements, which, in case of mobile or hardware wallets may be too demanding. At the same time, some of the operations carried by RGB Node (like consignment validation) doens't require access to the stash - while others can be performed just by accessing a disk-stored stash.

TL;DR: in most of the cases there is no need for a runtime which is constantly run in memory. A git-like solution working with disk storage in some directory may be enough to perform all the tasks which are needed for mobile, hardware wallets and, in many cases, desktop. 

The only situation where a constantly running runtime is needed is a server to do all tasks remotely - for instance to reduce consuming of mobile traffic in roaming when large consignments are involved they may be processed by a server. However this server may be a tiny layer of networking on top of the same library which is used on mobile.

## The poposal

Move most of the RGB Node functionality into RGB standard library and make it not requiring networking, permanent runtime, multithreading etc; giving much flexibility to wallet and software developers (like gitlib). Transform RGB node into a small server which uses this library and provides networking RPC API to it.

**Comment by cryptoquick** [comment]
This is an excellent idea, and it's basically proved out by the work we've done in bitmask-core. I'd love to be able to deprecate portions of the RGB sections if we can get the same functionality in the RGB std lib.

**Reply by dr-orlovsky** [reply]
What is "portions of RGB sections"? :)

**Reply by cryptoquick** [reply]
Much of our RGB code is lifted from the nodes and modified to run in-memory:
https://github.com/diba-io/bitmask-core/blob/development/src/operations/rgb/shared.rs

**Reply by dr-orlovsky** [reply]
I see... Now this sounds like "must do": even if we continue to use "bulk" RGB Node runtime on mobile (I see no point in doing that, but still) we need to move everything which is possible to the standard library out of it.

BTW, with the new node architecture I'm developing since Oct 2022 in @Cyphernet-WG we are finally getting rid of ZMQ, reduce number of threads and do most of the internal communications with `crossbeam_channel` right in-memory in WASM-compatible way. The most benefit of it will go to LNP Node (and BP Node), but still maybe it would help getting LN part into BitMask as well.

Actually, how in WASM you do file storage/DB? Through networking?

**Reply by cryptoquick** [reply]
That's great! Storage is currently being done outside the wasm binary, in the extension browser storage. We're currently working on Carbonado to improve that situation, though.

**Reply by dr-orlovsky** [reply]
When you accept a consignment, where it goes? Into the memory? And what if it would not fit?

Second: when you verify a consignment, how you access its data?

**Reply by cryptoquick** [reply]
Yes, in-memory, at least for now. Gigabyte consignments will be a long time off, I think, and even that would be fine on most computers.

**Reply by oacdutra** [reply]
This is an awesome idea, I think is good to helping the wallet developers to integrate and use RGB.

Can you think add this change in v0.10 milestone?

**Reply by dr-orlovsky** [reply]
yes, I'd like to get it planned for v0.10 - nut still need to hear what @zoedberg think of this idea

**Reply by zoedberg** [reply]
The idea in general is very good, though I don't think we should do this in version 0.10 since IMO that should contain only the last consensus breaking changes that are needed in order to stabilize the protocol and should have max priority. 

Moreover this will not totally solve the runtime issue (https://github.com/RGB-WG/rgb-node/issues/222) unless we also change the way stored works, since also stored is currently spawned as a thread.

> A git-like solution working with disk storage in some directory may be enough to perform all the tasks which are needed for mobile, hardware wallets and, in many cases, desktop.

I guess this is meant to substitute stored and that it's planned that libraries/wallet developers will implement their own storage. Did I understand correctly? Could you please explain how this should work? I'm wondering how rgb-std methods will instruct its callers which data needs to be saved.

Anyway, this seems like a lot of work and in order to facilitate upgrades I suggest not making too many changes in a single shot, a bug would probably be very hard to find in that case.

**Reply by cryptoquick** [reply]
All very good points, @zoedberg.

**Reply by oacdutra** [reply]
Yes, I agree with @zoedberg.

I think better move this to another milestone.

**Reply by dr-orlovsky** [reply]
@zoedberg thank you for all comments. Let me explain few points:

1. No, this will solve #222 since we get rid of stored or any threads at all (at least not controlled by a wallet)

2. No, this will not move storage thing to the wallet part.

3. No, this is not a huge work and actually is just a question of re-locating some code of RGB Node to RGB stdlib - and getting rids of most what is left in RGB Node (daemons, networking etc).

4. After this we would not need to release v0.10 of RGB Node - we will do just RGB stdlib v0.10 which will get what is needed from RGB Node. We can work on the rest of RGB Node later when there will be a need for it.

So in fact this reduces my work, not increases it.

**Reply by dr-orlovsky** [reply]
Trying to explain it in some other way.

Just see how `git` works: no server, no P2P protocol, just a directory which can grow up to gigabytes. In fact `git` would be a client-side-validated data if the history were commited into bitcoin as a single-use-seals. So `git` is a prototype of how RGB can work with client-side-validated data.

**Reply by zoedberg** [reply]
Ok so stored will not be replaced with another project, it will get deprecated, and for storage operations rgb-std methods will get a path to a data directory where a git-based storage will be used. Is my understanding correct now @dr-orlovsky? If so, I have some questions:

1. Why can't we keep using sled to save data?
2. Has a migration for existing nodes been considered? If so, has the time to write this migration been considered when you said that's "not a huge work"?

Talking instead about https://github.com/RGB-WG/rgb-node/issues/222, I think that in the long term it will still be necessary, because as soon as we'll find a solution for https://github.com/LNP-WG/lnp-node/discussions/71 we'll integrate storm and lnp-node runtimes in iris (via rgb-lib of course).

**Reply by dr-orlovsky** [reply]
> Ok so stored will not be replaced with another project, it will get deprecated, and for storage operations rgb-std methods will get a path to a data directory where a git-based storage will be used. Is my understanding correct now @dr-orlovsky?

Correct except that stored will not be "depricated" but just not used in RGB (but still can be used in other projects if needed).

> Why can't we keep using sled to save data?

I haven't said anything about the format of the on-disk data storage. This in fact can be anything - from what we have today (sled) to LLDB or anything else which we think would work. The considerations I have on this matter are:
1. I prefer not do write a lot of new storage-related code, so yes sled is the choice
2. It has to be not a C library but rather something rust native
3. It has to work well with WASM (?)
4. It has to work on mobile
5. It has to work on embeded platforms

> Has a migration for existing nodes been considered? If so, has the time to write this migration been considered when you said that's "not a huge work"?

I was assuming we continue using the same storage format, since I do not see clear reason to change it _at least in parallel with moving from RGB Node to git-like approach_. But if you see a reason to do that that has to be estimated separately, and it will not be related to the topic. Also it has to be better on the criteria 1-5 above than the current solution.

> Talking instead about https://github.com/RGB-WG/rgb-node/issues/222, I think that in the long term it will still be necessary

I am not getting what do you mean. What will be necessary? Stopping the thread? But there would be no threads in the library; all threads will be managed by a wallet. You need to stop it - you stop it.

**Reply by zoedberg** [reply]
> I was assuming we continue using the same storage format

Happy to hear this. Sorry for my misunderstading 

> I am not getting what do you mean. What will be necessary? Stopping the thread? But there would be no threads in the library; all threads will be managed by a wallet. You need to stop it - you stop it.

Yes, in the future we will need to stop spawned threads for lnp-node and storm-node (that, I guess for their nature must remain daemons). As discussed in https://github.com/RGB-WG/rgb-node/issues/222 and also in a dev call there is no simple way to stop a spawned thread in Rust and for that we will need a different thread managing architecture

**Reply by dr-orlovsky** [reply]
This is unrelated to the proposed changes to RGB Node, right?

As for LNP and Storm the new architecture is outthere already: check github.com/rust0amplify/io-reactor. It provides a way to shutdown the thread and launch it over again later. It is also capapble of managing thousands of connections on the same thread in a non-blocking way without use of rust async, i.e. much more lightweight

**Reply by zoedberg** [reply]
> This is unrelated to the proposed changes to RGB Node, right?

Yes, it's unrelated. Talking about proposed changes I'd say that if doing this refactor along with the consesus breaking changes helps reducing your work in a significant way I'm ok with it, if instead it's not so time-impacting I'd try to split work in 2 releases in order to facilitate wallets/libraries developers life

> As for LNP and Storm the new architecture is outthere already

Wow, this is great news! In which release do you think we're going to be able to change the runtime architecture in lnp-node and storm-node?

**Reply by dr-orlovsky** [reply]
> In which release do you think we're going to be able to change the runtime architecture in lnp-node and storm-node?

I'd like to see it happening in v0.10, which now for LNP/Storm would be able to happen independently from RGB release (i.e. later).

IO reactor is production-ready (it was tested even outside of LNP/BP projects), so it is just a question of switching to it.
